lib/datasource.js

```diff
@@ -2115,26 +2157,116 @@ DataSource.prototype.transaction = function() {
 
   transaction.isTransaction = true;
   transaction.origin = dataSource;
-  transaction.name = dataSource.name;
-  transaction.settings = dataSource.settings;
   transaction.connected = false;
   transaction.connecting = false;
-  transaction.connector = dataSource.connector.transaction();
 
   // create blank models pool
-  transaction.modelBuilder = new ModelBuilder();
-  transaction.models = transaction.modelBuilder.models;
-  transaction.definitions = transaction.modelBuilder.definitions;
+  var modelBuilder = new ModelBuilder();
+  var slaveModels = modelBuilder.models;
+  transaction.modelBuilder = modelBuilder;
+  transaction.models = slaveModels;
+  transaction.definitions = modelBuilder.definitions;
+
+  // For performance reasons, use a getter per model and only call copyModel()
+  // for the models that are actually used. These getters are then replaced
+  // with the actual values on first use.
+  var masterModels = dataSource.modelBuilder.models;
+  Object.keys(masterModels).forEach(function(name) {
+    Object.defineProperty(slaveModels, name, {
+      enumerable: true,
+      configurable: true,
+      get: function() {
+        // Delete getter so copyModel() can redefine slaveModels[name].
+        delete slaveModels[name];
+        return dataSource.copyModel.call(transaction, masterModels[name]);
+      },
+    });
+  });
+
+  function handleExecute() {
+    if (execute) {
+      cb = cb || utils.createPromiseCallback();
+      var done = function(err) {
+        if (err) {
+          transaction.rollback(function(error) {
+            cb(err || error);
+          });
+        } else {
+          transaction.commit(cb);
+        }
+        // Make sure cb() isn't called twice, e.g. if `execute` returns a
+        // thenable object and also calls the passed `cb` function.
+        done = function() {};
+      };
+      try {
+        var result = execute(slaveModels, done);
+        if (result && typeof result.then === 'function') {
+          result.then(function() { done(); }, done);
+        } else if (execute.length < 2) {
+          // execute didn't return a thenable and doesn't receive a callback.
+          // It must have executed synchronously, so call done() now.
```

Comment:
Is this a safe assumption? What if the function to be executed is reading `arguments` directly? Is it necessary to support the use case of synchronous `execute`?

Reply:
I've replicated this from the scripts in `loopback-boot`... I don't think you'd read a `done` method from arguments lists? LB already makes these kind of assumptions in lots of places.

Reply:
I see. I personally prefer to keep our API as small as possible and if we don't need to support sync `execute` for backwards compatibility, then I'd rather leave the support out. The less features/code lines we have to maintain, the easier for us :)

Reply:
Having said that, this is a detail and I am fine with going with what you have proposed here.

Reply:
Ok, I agree! It doesn't make much sense to have sync code here anyway. I'll take it out.

Reply:
I've addressed your feedback in 2ad198278b62aa988d0e66b80638ec970e7774cf